presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,406 @@
+            return Optional.empty();
+        }
+
+        JoinEnumerationResult result = new JoinEnumerator(idAllocator, symbolAllocator, session, lookup, multiJoinNode.getFilter(), costComparator).chooseJoinOrder(multiJoinNode.getSources(), multiJoinNode.getOutputSymbols());
+        return result.getCost().isUnknown() || result.getCost().equals(INFINITE_COST) ? Optional.empty() : result.getPlanNode();
+    }
+
+    @VisibleForTesting
+    static class JoinEnumerator
+    {
+        private final Map<Set<PlanNode>, JoinEnumerationResult> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final Ordering<JoinEnumerationResult> resultOrdering;
+        private final EqualityInference allInference;
+        private final Expression allFilter;
+        private final SymbolAllocator symbolAllocator;
+        private final Lookup lookup;
+
+        @VisibleForTesting
+        JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup, Expression filter, CostComparator costComparator)
+        {
+            requireNonNull(idAllocator, "idAllocator is null");
+            requireNonNull(symbolAllocator, "symbolAllocator is null");
+            requireNonNull(session, "session is null");
+            requireNonNull(lookup, "lookup is null");
+            requireNonNull(filter, "filter is null");
+            requireNonNull(costComparator, "costComparator is null");
+            this.idAllocator = idAllocator;
+            this.symbolAllocator = symbolAllocator;
+            this.session = session;
+            this.lookup = lookup;
+            this.resultOrdering = getResultOrdering(costComparator, session);
+            this.allInference = createEqualityInference(filter);
+            this.allFilter = filter;
+        }
+
+        private static Ordering<JoinEnumerationResult> getResultOrdering(CostComparator costComparator, Session session)
+        {
+            return new Ordering<JoinEnumerationResult>()
+            {
+                @Override
+                public int compare(JoinEnumerationResult result1, JoinEnumerationResult result2)
+                {
+                    return costComparator.compare(session, result1.cost, result2.cost);
+                }
+            };
+        }
+
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
+            JoinEnumerationResult bestResult = memo.get(multiJoinKey);
+            if (bestResult == null) {
+                checkState(sources.size() > 1, "sources size is less than or equal to one");
+                ImmutableList.Builder<JoinEnumerationResult> resultBuilder = ImmutableList.builder();
+                Set<Set<Integer>> partitions = generatePartitions(sources.size()).collect(toImmutableSet());
```

Comment:
IIUC, we could possibly spare lots of object allocations if iterating over partitions based on BitSets instead of `Set<Integer>`. However, we would need something like https://gist.github.com/avianey/3ea9dbc64489e4058d422d3b270644a0 .